### PR TITLE
support for math functions(special.erfinv, special.erfcinv, math.frex…

### PR DESCRIPTION
**Pull request by Marojoh**
## What this PR is about
* support for math functions special.erfinv, special.erfcinv, math.frexp, math.log2, math.modf, linalg.norm, math.ldexp, math.isfinite

## Tests
*  Added some test cases to test_math.py for special.erfinv, special.erfcinv, math.log2.
* Past tests
* Performed personal tests for all added math functions above, and they all past

## TODOs 

* Support test cases for the others in test_math.py will be appreciated.

### CODE REVIEW

File: numba/cuda/libdevice.py [hunk 1]
```diff
@@ -78,11 +83,17 @@ def core(context, builder, sig, args):
...
+unarys += [('__nv_erfcinv', '__nv_erfcinvf', special.erfcinv)]
 unarys += [('__nv_tgamma', '__nv_tgammaf', math.gamma)]
 unarys += [('__nv_lgamma', '__nv_lgammaf', math.lgamma)]
+unarys += [('__nv_modf', '__nv_modff', math.modf)]
+unarys += [('__nv_frexp', '__nv_frexpf', math.frexp)]
+unarys += [('__nv_norm', '__nv_normf', linalg.norm)]
 unarys += [('__nv_sqrt', '__nv_sqrtf', math.sqrt)]
 unarys += [('__nv_log', '__nv_logf', math.log)]
 unarys += [('__nv_log10', '__nv_log10f', math.log10)]
+unarys += [('__nv_log2', '__nv_log2f', math.log2)]
```
Comment (stuartarchibald) [inline numba/cuda/libdevice.py]: This will need a version guard, it was added in Python 3.3.

File: numba/cuda/libdevice.py [hunk 2]
```diff
@@ -101,6 +112,7 @@ def core(context, builder, sig, args):
 binarys += [('__nv_copysign', '__nv_copysignf', math.copysign)]
 binarys += [('__nv_atan2', '__nv_atan2f', math.atan2)]
 binarys += [('__nv_pow', '__nv_powf', math.pow)]
+binarys += [('__nv_ldexp', '__nv_ldexpf', math.ldexp)]
```
Comment (stuartarchibald) [inline numba/cuda/libdevice.py]: The second argument it integer, so adding it here would not be valid as the lowering implementation assumes floats:
https://github.com/numba/numba/pull/4855/files#diff-063ba8342a9754fcd66d73729ca1a1bdR134-R137

I expect a custom lowering function will be needed.

File: numba/cuda/cudamath.py [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 import math
-from numba import types, utils
-from numba.typing.templates import (AttributeTemplate, ConcreteTemplate,
-                                    signature, Registry)
+from scipy import special
+from scipy import linalg
+from numba.extending import overload
+from numba import (
+     cuda, types, utils)
+from numba.typing.templates import (
```
Comment (stuartarchibald) [inline numba/cuda/cudamath.py]: Please do not reformat existing code.

File: numba/cuda/cudamath.py [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
+    AttributeTemplate, ConcreteTemplate,
+    signature, Registry)
+#from numba.errors import TypingError
```
Comment (stuartarchibald) [inline numba/cuda/cudamath.py]: Dead comment, please remove.

File: numba/cuda/tests/cudapy/test_math.py [hunk 5]
```diff
@@ -410,7 +472,7 @@ def ufunc(x):
         self.unary_template_float32(math_lgamma, ufunc, start=0.1)
         self.unary_template_float64(math_lgamma, ufunc, start=0.1)
 
-    #------------------------------------------------------------------------------
+    #---------------------------------------------------------------------------------
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_math.py]: Is this (and the following) change absolutely necessary?

**Comment by Marojoh** [comment]
Also on the issue of math.log2, what do you mean by version guard?

File: numba/cuda/cudamath.py [hunk 6]
```diff
@@ -1,13 +1,18 @@
 from __future__ import print_function, absolute_import, division
 import math
-from numba import types, utils
-from numba.typing.templates import (AttributeTemplate, ConcreteTemplate,
-                                    signature, Registry)
+from scipy import special
```
[the remark this answers is not in the document]
Comment (Marojoh) [reply numba/cuda/cudamath.py]: following up on my response yesterday, on the situation of the scipy dependency.  What solution will you propose in solving this problem?

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cuda/cudamath.py]: You can add this locally in your own code, or, alternatively, open a ticket to discuss on https://github.com/numba/numba-scipy (which is a new project to support SciPy functions in Numba).

**Comment by gmarkall** [comment]
@Marojoh As Numba can't take a dependency on SciPy, the way forward for the parts of your PR that implement functions from the `math` module would be to make a PR for them without any of the SciPy functions.

For the parts that implement SciPy functions, opening a PR or issue in https://github.com/numba/numba-scipy to discuss the way forward there would be appropriate - the purpose of numba-scipy is explicitly to extend Numba to support SciPy functions, so it would be great to find a way to get additional CUDA target implementations of SciPy functions into it.

**Comment by stuartarchibald** [comment]
Just checking in on the state of this PR. @Marojoh was wondering if you have had time to look the suggestions above? Closing this PR unmerged is another option if it's irretrievably stalled/better placed in e.g. `numba-scipy`. Thanks.

**Comment by stuartarchibald** [comment]
I'm going to close this as stalled due to numerous unresolved items. Please do re-open or forward patches to a new PR if work continues. Many thanks.

**Pull request closed without merging**